### PR TITLE
Add interfaces CSV import, interfaces list, and import/export buttons to device

### LINKED ISSUE #1347: Bulk CSV import for more object types
### Issue type: feature request

**NetBox version:** 2.0.9

I'm trying to migrate to NetBox from a legacy system (Racktables) with quite a bit of data in it (about 3000 pieces of equipment, in multiple racks/rows/sites/regions).

While looking into how to import the data, I noticed that while it's possible to import via CSV, the functionality is limited to a few specific object types (e.g. sites, tenants, racks, devices) and no such functionality exists for the others. I'm missing primarily manufacturers and device types (I have 86 models from various vendors) and rack groups (I have 15), but it'd be neat to have some of the others as well (regions, platforms, device roles).

Are there any plans to implement such a thing? Should I be looking at alternatives like e.g. using the API instead?

Note that #451 is a similar ticket but a) limited to device types and b) also trying to address the orthogonal problem of sharing a common base for device types between users. Finally, for my trial, I'm also missing the capability of importing custom fields, which is tracked separately as #568.

### PR DESCRIPTION
**Pull request by netsiphon**
Fixes: #1347 Bulk CSV import for more object types

If you would prefer I can open a feature request/requests specifically for these changes under my name and resubmit. Apologize for all of the tiny commits...netbox dev running on non-local vm and I'm a Django novice.

Proposed changes:

1. Add Import Interfaces CSV form
2. Add Interfaces List form to allow search and filtering for export
3. Add Import and Export Interfaces buttons to Device view with direct link to Interface List form export(for specific device).

### CODE REVIEW

**Comment by netsiphon** [comment]
Besides resolving the conflicts with base.html can I do anything else to make these changes includable? Is it a matter of project roadmap or my implementation?

File: netbox/dcim/constants.py [hunk 1]
```diff
@@ -48,6 +48,19 @@
     [IFACE_ORDERING_NAME, 'Name (alphabetically)']
 ]
 
+# Interface enabled as choice
+IFACE_DISABLED = False
```
Comment (jeremystretch) [inline netbox/dcim/constants.py]: `enabled` is a boolean field; no need for constants.

File: netbox/dcim/filters.py [hunk 2]
```diff
@@ -577,6 +577,94 @@ def _mac_address(self, queryset, name, value):
             return queryset.none()
 
 
+class InterfaceListFilter(django_filters.FilterSet):
```
Comment (jeremystretch) [inline netbox/dcim/filters.py]: We already have an `InterfaceFilter` class.

File: netbox/dcim/forms.py [hunk 3]
```diff
@@ -1531,6 +1531,114 @@ class InterfaceBulkDisconnectForm(ConfirmationForm):
...
+    )
+    mtu = forms.IntegerField(
+        required=False,
+        help_text='MTU'
+    )
+    mgmt_only = forms.CharField(
+        required=False,
+        help_text='Management Only'
+    )
+    is_virtual = forms.BooleanField(
```
Comment (jeremystretch) [inline netbox/dcim/forms.py]: None of the `is_*` properties are database fields.

File: netbox/dcim/urls.py [hunk 4]
```diff
@@ -177,6 +177,8 @@
     url(r'^interface-connections/(?P<pk>\d+)/delete/$', views.interfaceconnection_delete, name='interfaceconnection_delete'),
     url(r'^interfaces/(?P<pk>\d+)/edit/$', views.InterfaceEditView.as_view(), name='interface_edit'),
     url(r'^interfaces/(?P<pk>\d+)/delete/$', views.InterfaceDeleteView.as_view(), name='interface_delete'),
+    url(r'^interfaces/$', views.InterfaceListView.as_view(), name='interface_list'),
```
Comment (jeremystretch) [inline netbox/dcim/urls.py]: There's no need for a view which exposes device components outside the context of a device.

**Review by jeremystretch — CHANGES_REQUESTED** [review]
I appreciate the effort, but there's a lot of stuff in here that needs to be reworked or simply isn't needed. Primarily, we don't want to introduce a view which lists interfaces on their own; like other components, interfaces should normally only be shown within the context of their parent device. Also, please review the data model for interfaces: You've introduced several fields in the import form which do not correspond to database fields.

**Comment by netsiphon** [comment]
Sorry I haven't been able to get back to this...I think before I rework the changes, I'll have to convince you of the need for an interfaces view if only for use in filtering interface exports and also for searching through interface information all in one place.  

I believe you said elsewhere that the description field is populated from ifDescr through napalm, but we've attempted to use it for label information when it comes to patch panels far end connections.

I suppose we could define wall plates as devices and give them interfaces, then make the interface connections but that will create a tremendous number of devices and still require us to export interfaces for multiple devices at the same time. Also, it would be good to be able to populate a db field for the ifAlias field as we often try to put far end connections into the device configuration but that could be for another day.

**Comment by jeremystretch** [comment]
> I believe you said elsewhere that the description field is populated from ifDescr through napalm

NetBox does not populate anything via NAPALM. NAPALM is only used to validate live data against what has been administratively defined in NetBox.

> but we've attempted to use it for label information when it comes to patch panels far end connections

NetBox doesn't yet support modeling a physical cable plant (see #20).

**Comment by netsiphon** [comment]
Ok, keeping physical devices out of it as I see the disclaimer.

So you would incorporate the interface import view(with fixes) allowing import from any number of devices but export only if it is limited to the single device view alone yes? Is the search or export of interfaces based on a filter something you would want to limit to a report or backend db access?

I ask because we still have chassis devices with 500+ interfaces so even visually searching within the same device isn't necessarily quick.

**Pull request closed without merging**